Add `nicename_exists` function and associated tests

pbearne opened the pull request:
The `nicename_exists` function is implemented to determine if a given user_nicename exists. The function is also integrated into the user nickname checking mechanism within the user class. Additionally, various test cases for the new function have been added, which cover scenarios with existing, non-existing, and nicknames associated with different user logins.

Trac ticket: https://core.trac.wordpress.org/ticket/44921

Commit:
Add `nicename_exists` function and associated tests

The `nicename_exists` function is implemented to determine if a given user_nicename exists. The function is also integrated into the user nickname checking mechanism within the user class. Additionally, various test cases for the new function have been added, which cover scenarios with existing, non-existing, and nicknames associated with different user logins.

Commit:
Add `nicename_exists` function and associated tests

The `nicename_exists` function is implemented to determine if a given user_nicename exists. The function is also integrated into the user nickname checking mechanism within the user class. Additionally, various test cases for the new function have been added, which cover scenarios with existing, non-existing, and nicknames associated with different user logins.

mukeshpanchal27 commented on src/wp-includes/user.php:
```diff
@@ -1988,6 +1988,32 @@ function username_exists( $username ) {
+ * @param string $nicename Nicename.
+ * @param string $login (Optional) User login to compare against.
+ *
+ * @return int|false The user's ID on success, and false on failure.
```
```suggestion
 * @param string $nicename Nicename.
 * @param string $login    Optional. User login to compare against.
 * @return int|false The user ID on success, false on failure.
```

Docblock suggestion.

mukeshpanchal27 commented on src/wp-includes/user.php:
```diff
@@ -1988,6 +1988,32 @@ function username_exists( $username ) {
+ * the {@link https://developer.wordpress.org/themes/basics/conditional-tags/
+ * Conditional Tags} article in the Theme Developer Handbook.
+ *
+ * @since 6.6.0
```
```suggestion
 * @since 6.7.0
```

mukeshpanchal27 commented on src/wp-includes/user.php:
```diff
@@ -1988,6 +1988,32 @@ function username_exists( $username ) {
 }
 
+/**
+ * Determines whether the given user_nicename exists.
+ * For more information on this and similar theme functions, check out
```
```suggestion
 * Determines whether the given user_nicename exists.
 *
 * For more information on this and similar theme functions, check out
```

mukeshpanchal27 commented on src/wp-includes/user.php:
```diff
@@ -1988,6 +1988,32 @@ function username_exists( $username ) {
+function nicename_exists( $nicename, $login = false ) {
+	$user_id = false;
+	$user = get_user_by( 'slug', $nicename );
+	if ( $user ) {
+		if ( $login !== $user->user_login ) {
+			$user_id = $user->ID;
+		}
+	}
```
```suggestion
	if ( $user && $login !== $user->user_login ) {
		$user_id = $user->ID;
	}
```

mukeshpanchal27 commented on src/wp-includes/user.php:
```diff
@@ -1988,6 +1988,32 @@ function username_exists( $username ) {
+	return $user_id;
+}
+
+
```
```suggestion
```

mukeshpanchal27 commented on src/wp-includes/user.php:
```diff
@@ -1988,6 +1988,32 @@ function username_exists( $username ) {
+ *
+ * @return int|false The user's ID on success, and false on failure.
+ */
+function nicename_exists( $nicename, $login = false ) {
```
Could we use function name as `user_nicename_exists` as it more clear then existing?

mukeshpanchal27 commented on tests/phpunit/tests/user/nicenameExisits.php:
```diff
@@ -0,0 +1,56 @@
+ *
+ * @covers ::nicename_exists
+ */
+class Tests_User_NicenameExists extends WP_UnitTestCase {
```
```suggestion
class Tests_User_Nicename_Exists extends WP_UnitTestCase {
```

mukeshpanchal27 reviewed (changes requested):
@pbearne Left some feedback. Look solid to me.

pbearne committed:
Update src/wp-includes/user.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

pbearne committed:
Update src/wp-includes/user.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

pbearne committed:
Update src/wp-includes/user.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

pbearne committed:
Update src/wp-includes/user.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

pbearne committed:
Update src/wp-includes/user.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

pbearne committed:
Update tests/phpunit/tests/user/nicenameExisits.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

Commit:
Rename `nicename_exists` to `user_nicename_exists`.

Updated function and test references to improve clarity and consistency. This change ensures alignment with naming conventions and enhances readability within the codebase.

mukeshpanchal27 commented on src/wp-includes/user.php:
```diff
@@ -1988,6 +1988,29 @@ function username_exists( $username ) {
+ * @param string $login    Optional. User login to compare against.
+ * @return int|false The user ID on success, false on failure.
+ */
+function user_nicename_exists($nicename, $login = false ) {
```
```suggestion
function user_nicename_exists( $nicename, $login = false ) {
```

PHPCS fix

mukeshpanchal27 commented on src/wp-includes/user.php:
```diff
@@ -1988,6 +1988,29 @@ function username_exists( $username ) {
+ * @return int|false The user ID on success, false on failure.
+ */
+function user_nicename_exists($nicename, $login = false ) {
+	$user_id = false;
+	$user = get_user_by( 'slug', $nicename );
```
```suggestion
	$user_id = false;
	$user    = get_user_by( 'slug', $nicename );
```

pbearne committed:
Update src/wp-includes/user.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

pbearne committed:
Update src/wp-includes/user.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

mukeshpanchal27 commented on tests/phpunit/tests/user/nicenameExisits.php:
```diff
@@ -0,0 +1,56 @@
+<?php
+
+/**
+ * Tests the `nicename_exists` function.
```
```suggestion
 * Tests the `user_nicename_exists` function.
```

mukeshpanchal27 commented on tests/phpunit/tests/user/nicenameExisits.php:
```diff
@@ -0,0 +1,56 @@
+class Tests_User_Nicename_Exists extends WP_UnitTestCase {
+
+	/**
+	 * Tests that `nicename_exists` returns the user ID when the nicename exists.
```
```suggestion
	 * Tests that `user_nicename_exists` returns the user ID when the nicename exists.
```

mukeshpanchal27 commented on tests/phpunit/tests/user/nicenameExisits.php:
```diff
@@ -0,0 +1,56 @@
+	 *
+	 * @ticket 44921
+	 */
+	public function test_nicename_exists_with_existing_nicename() {
```
```suggestion
	public function test_user_nicename_exists_with_existing_nicename() {
```

mukeshpanchal27 commented on tests/phpunit/tests/user/nicenameExisits.php:
```diff
@@ -0,0 +1,56 @@
+	}
+
+	/**
+	 * Tests that `nicename_exists` returns false when the nicename does not exist.
```
```suggestion
	 * Tests that `user_nicename_exists` returns false when the nicename does not exist.
```

mukeshpanchal27 commented on tests/phpunit/tests/user/nicenameExisits.php:
```diff
@@ -0,0 +1,56 @@
+	 *
+	 * @ticket 44921
+	 */
+	public function test_nicename_exists_with_nonexistent_nicename() {
```
```suggestion
	public function test_user_nicename_exists_with_nonexistent_nicename() {
```

mukeshpanchal27 commented on tests/phpunit/tests/user/nicenameExisits.php:
```diff
@@ -0,0 +1,56 @@
+	}
+
+	/**
+	 * Tests that `nicename_exists` returns false when the nicename exists but belongs to a different user.
```
```suggestion
	 * Tests that `user_nicename_exists` returns false when the nicename exists but belongs to a different user.
```

mukeshpanchal27 commented on tests/phpunit/tests/user/nicenameExisits.php:
```diff
@@ -0,0 +1,56 @@
+	 *
+	 * @ticket 44921
+	 */
+	public function test_nicename_exists_with_different_user_login() {
```
```suggestion
	public function test_user_nicename_exists_with_different_user_login() {
```

mukeshpanchal27 commented on tests/phpunit/tests/user/nicenameExisits.php:
```diff
@@ -0,0 +1,56 @@
+	}
+
+	/**
+	 * Tests that `nicename_exists` returns false when the nicename exists but belongs to a different user.
```
```suggestion
	 * Tests that `user_nicename_exists` returns false when the nicename exists but belongs to a different user.
```

mukeshpanchal27 commented on tests/phpunit/tests/user/nicenameExisits.php:
```diff
@@ -0,0 +1,56 @@
+	 *
+	 * @ticket 44921
+	 */
+	public function test_nicename_exists_with_same_user_login() {
```
```suggestion
	public function test_user_nicename_exists_with_same_user_login() {
```

mukeshpanchal27 reviewed (approved):
Thanks @pbearne for the update.

Pre-approved. Left some minor non blocking suggestions also update unit test file name to `userNicenameExisits.php`

mukeshpanchal27 commented:
@peterwilsoncc If you get chance please review it as it's so close.

pbearne committed:
Update tests/phpunit/tests/user/nicenameExisits.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

pbearne committed:
Update tests/phpunit/tests/user/nicenameExisits.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

pbearne committed:
Update tests/phpunit/tests/user/nicenameExisits.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

pbearne committed:
Update tests/phpunit/tests/user/nicenameExisits.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

pbearne committed:
Update tests/phpunit/tests/user/nicenameExisits.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

pbearne committed:
Update tests/phpunit/tests/user/nicenameExisits.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

pbearne committed:
Update tests/phpunit/tests/user/nicenameExisits.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

pbearne committed:
Update tests/phpunit/tests/user/nicenameExisits.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

pbearne committed:
Update tests/phpunit/tests/user/nicenameExisits.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

peterwilsoncc commented on src/wp-includes/user.php:
```diff
@@ -1988,6 +1988,29 @@ function username_exists( $username ) {
+ * @param string $login    Optional. User login to compare against.
+ * @return int|false The user ID on success, false on failure.
+ */
+function user_nicename_exists( $nicename, $login = false ) {
```
I'd name it `wp_unique_user_nicename()` to match the same for term and post slugs. If the name exists then it can return a version with a suffix.

mukeshpanchal27 replied:
+1 for name changes.

spacedmonkey reviewed (approved):
LGTM

peterwilsoncc reviewed (commented):
Forgot to submit my review lst week.

I've posted the naming things suggestion on the ticket but am also after some clarification about how this is a real world improvement, and why querying the full user objects rather than simply the nice names helps.

See [WP-44921 (comment)](https://core.trac.wordpress.org/ticket/44921#comment:16)

mukeshpanchal27 commented:
I did some benchmarking for current trunk and the PR changes with 100 and 1000 user inserts and the PR didn't show any performance regression instead it improve the performance if we use same nicename.

Result:
```
Trunk: Total time to insert 100 users: 2.0517628192902 seconds.
PR: Total time to insert 100 users: 1.4436409473419 seconds.

Trunk: Total time to insert 1000 users: 80.944605112076 seconds.
PR: Total time to insert 1000 users: 16.288486003876 seconds.
```

Could anyone of you please benchmarking your end?

peterwilsoncc commented:
> I did some benchmarking for current trunk and the PR changes with 100 and 1000 user inserts and the PR didn't show any performance regression instead it improve the performance if we use same nicename.

This isn't the question I am asking on the trac ticket. I am asking what the real world improvements are.

I accept that this change will improve performance of inserting 1000 users with the identical nice name `test_user` but that doesn't reflect the real world situation in which people are likely to use their actual name for the nice name.

Even common names such as John or Peter aren't likely to result in the need to iterate thought 100 different user accounts to find a suitable suffix.

mukeshpanchal27 commented:
Got it but for bigger user base site have more then 100 same nicename so i would be useful there.
